### ISSUE TITLE
Add suspend and resume functions

### ISSUE THREAD

**Issue by malchmih** [tolitius/mount#5]
I'd like to suggest adding `:resume` and `:suspend` functions to `defstate`. There is a library that does that for Component: https://github.com/weavejester/suspendable. From my experience, it is very convenient during development.

**Comment by tolitius** [comment]
@malchmih, great idea.

For the use case defined in https://github.com/weavejester/suspendable, I had to [pass a `0` port to Jetty](https://github.com/tolitius/mount/blob/34bdb4dbe76897770e5a2052835e17ec156d6ca7/doc/uberjar.md#repl-time) in development, so it just binds on the random port on each `(reset)`.

But I agree, making it `unstoppable` via `(reset)` while `stoppable` via `(mount/stop)` is useful, which needs an additional pair of actions like `suspend`/`resume`.

---

This needs some thinking though, since
- it would complicate `defstate` a bit:

``` clojure
(defstate www :start (start-server)
              :stop (.stop www))
```

vs.

``` clojure
(defstate www :start (start-server)
              :stop (.stop www)
              :suspend :do-nothing
              :resume (restart-if-changed conf))  ;;  where conf would have a port
```

`:suspend` here can be omitted of course.
- most of the states won't need :suspend/:resume

which means that in development 

_either_: for a single `(reset)` to support `:suspend/:resume`, _all states_ would have to implement `:suspend`/`:resume` in order to support the need of _only few_ (or.. mount can check if there is no `suspend` call `stop`, no `resume`, call `start`)

_or_: using`(reset)` for most of the states and using `(resume)`/`(suspend)` for some.

Where I do see it useful is ironically not in dev mode, but in production: when a server needs to be taken out of rotation in a datacenter, it's (for example) queue listeners or http endpoints can be suspended, and then resumed right before it goes back being active.

It might be solved with [starting / stopping only certain states](https://github.com/tolitius/mount#start-and-stop-parts-of-application) (a part of an application). But the `:suspend`/`:resume` idea is definitely worth thinking about.

**Comment by malchmih** [comment]
@tolitius In case of the original library

> If a component does not explicitly satisfy Suspendable, it will be stopped and started as normal.

which is a sane fallback and could be applied here too. So users will add `:suspend/:resume` only where they need them.

Also here you can find some examples of suspendable components: https://github.com/weavejester/duct/wiki/Compatible-Libraries. And I can provide some more if you need.

**Comment by tolitius** [comment]
agree. I added `or.. mount can check if there is no suspend call stop, no resume, call start` to the original reply a couple of seconds after commenting.

The nice thing about suspendable is that [it is a protocol](https://github.com/weavejester/suspendable/blob/master/src/suspendable/core.clj#L7), so a component that wants to be either "suspended and resumed", or "not bothered and resumed if needed" _has to_ implement both methods. Even if one of them does nothing:

``` clojure
Suspendable
  (suspend [component]
    component)
  (resume [component old-component]
    (if (= (:port component) (:port old-component))
      ...)))
```

In mount this can be a little strange to require `:suspend :do-nothing` for components that only want to restart if something changed:

``` clojure
(defstate www :start (start-server)
              :stop (.stop www)
              :suspend :do-nothing
              :resume (restart-if-changed conf))
```

If we omit the `:suspend :do-nothing`, it leaves `www` in a strange "state" (no pun intended). Should it be stopped on `(mount/suspend)`? Should it be started when a `(mount/start)` is called, but it is currently `:suspended`?

What do you think?

**Comment by malchmih** [comment]
> Should it be stopped on (mount/suspend)? Should it be started when a (mount/start) is called, but it is currently :suspended?

Do you mean from the user's perspective? `mount` will know what to do, right? If I understood correctly, then we could document the default and I would be up to user what to do on `resume`.

**Comment by tolitius** [comment]
I mean the developer's expectations and solid defaults. I'd like them to match :)

As to omitting `:suspend`:

``` clojure
(defstate www :start (start-server)
              :stop (.stop www)
              :resume (restart-if-changed conf))
```

most likely a default behavior in this case is to: 
- treat any state that has `:resume` as `suspendable`
- make sure `:suspend` cannot exist without `:resume`

---

In apps with at least one state that needs to be suspended/resumed, while `(mount/suspend)` / `(mount/resume)` will work, it is not clear what `(mount/start)` / `(mount/stop)` should do, since both may end up caring about:
- Am I starting this state from a clean state or from a suspended state?
- Am I stopping this state from a started state or from a suspended state?

The ultimate solutions / defaults should allow all 4 `start`/`stop`/`suspend`/`resume` to coexist peacefully.
In order to have deterministic states on start/stop, we can probably flip the behavior to:
- on start => if a state is `:suspended`, call resume
- on stop => if a state is `:suspended`, call resume then stop

not exactly in love with `resuming on stop` in order to stop it (definitely incorrect :)), needs more thinking.

**Comment by malchmih** [comment]
Ah, I see now, you care too much about users =)
In Component nothing stops you from calling `start` twice in a row. And it may lead to memory leaks, exceptions, etc. So it is ultimately user's responsibility to make it work correctly. That's why there is an [Idempotence section](https://github.com/stuartsierra/component#idempotence). But also user know, that `reset` won't do that - it will call `start` only after `stop`.
In case of `state` with `resume/suspend`, these functions will be called by `reset`. Is a user decides to call `start` on `suspended`, then blame is on him and he should make it work.
But now I notice that you keep state internally (`started?`). From my point of view, you are doing too much. What I love about Component is it's simplicity and transparency. If it was my decision to make, I'd go with allowing simple access to an internal state `:started?/suspended?` and letting users of the library manage the rest of it themselves.
In any case, everything should be explicitly stated in the documentation with "Best Practices" section and some examples.
I hope you understand my position on the subject here =)

**Comment by tolitius** [comment]
@malchmih yep, your position is clear and I appreciate you sharing it.

I do see why making mount 100% transparent, relying on users to read the docs and care about start/stop idempotence themselves feels right. but.. :)

The are a couple of reasons I chose to teach mount how to avoid double starts / stops:
- Good default

I can't think of a use case where a developer would want to double start or double stop (e.g. to `start` something more then once without stopping (suspending) it beforehand). I am not saying there are no such use cases, but I can't think of any, can you?
- Selfish

It is a really useful behavior (to me) :) Which tells me it might be for other people.

As to `:suspended?` state I like your approach with allowing to check "the state of the state". But I need to think a little more on how to do that, because start/stop/suspend/resume functions know nothing about mount (which I really like), but also mount only knows states by `#'full.namespace/state`. 

It's not _really_ simple to `(mount/check-state #'full.namespace/state)`. Also the same start/stop functions can be used to start/stop _different_ states (e.g. a database connection with different schemas, etc.), and it would not be clean to couple their implementations with concrete states.

**Comment by malchmih** [comment]
> I can't think of a use case where a developer would want to double start or double stop (e.g. to start something more then once without stopping (suspending) it beforehand). I am not saying there are no such use cases, but I can't think of any, can you?

No, I can't and I'm 100% agree with you here :blush: 

> It is a really useful behavior (to me) :) Which tells me it might be for other people.

Absolutely, :+1: Don't get me wrong, I really love libraries that simplify user's life. I'm just trying to avoid maintenance problems in the future, since we unlikely to predict all possible use cases. 

But this approach doesn't work for suspend/resume in my opinion. 90-95% use cases for suspendable `states` will look like this:

``` clojure
:start (real start here)
:stop (real stop)
:suspend :do-nothing
:resume (if (= new-args old-args) 
                  :do-nothing 
                  (do (stop) (start)))
```

and you can definitely invent some defaults here.
But another 5% will be like [this one](https://github.com/weavejester/ring-jetty-component/blob/master/src/ring/component/jetty.clj), which does "real stop" on `stop`, "real start" on `start` and suspends/resumes requests to a handler on `suspend/resume`. This example naturally fits in the 4-state model, right? For this to work it should know it's current state.
Just stated problems here, will need more time to think also =)

**Comment by malchmih** [comment]
If we assume (and document for users) that `stop` stops for good - releases all resources even for suspended state, then the table below probably makes sense.

| State\Function | `start` | `stop` | `suspend` | `resume` |
| --- | --- | --- | --- | --- |
| stopped | ok | noop | noop | start |
| started | won't happen | ok | ok | noop |
| suspended | resume | **stop** | noop | ok |

What do you think?

**Comment by tolitius** [comment]
>  stop stops for good - releases all resources even for suspended state

I like it. The table looks solid.

I am currently at Datomic conference (a pre conference to [Clojure/Conj](http://clojure-conj.org/)), and @stuarthalloway is rocking it with Datomic intel. Which is great, but I can't focus at the moment :) to find an answer to this question:
- Can you think of any use cases, where simply stopping a `:suspended` state on `stop` is not possible / clean / desired?

I also think, there should be a way to _optionally_ change defaults. e.g if someone needs to alter the mount flow for a certain app, would be good to make it simple to do so. This is not at all a stopper for `suspend`/`resume` feature, but something to think about.

**Comment by malchmih** [comment]
Sure, going into hammock state now :smile: Good luck with the conference!

**Comment by malchmih** [comment]
> Can you think of any use cases, where simply stopping a :suspended state on stop is not possible / clean / desired?

No, I can't. From my point of view suspended state is a subset of the running state in terms of resources (not sure how to explain it better). So `stop` should work here.
Also `resume/suspend` are only there for developer's convenience, right? If you are struggling to make it work, just don't use it - it won't be used in production anyway.

> I also think, there should be a way to optionally change defaults.

Not sure what exactly you mean by that. Do you have some example in mind?

**Comment by tolitius** [comment]
sounds good. then we'll go with your idea:

| State\Function | `start` | `stop` | `suspend` | `resume` |
| --- | --- | --- | --- | --- |
| stopped | ok | noop | noop | start |
| started | won't happen | ok | ok | noop |
| suspended | resume | **stop** | noop | ok |

---

as far as changing defaults, I meant something like lein/boot props, i.e.:

``` clojure
  :profiles {:dev {:mount {:stop-on-suspend? false
                           :exclude ["#'app.test" "#'app.qa"]
                           :and {:something "else"}}}}
```

**Comment by malchmih** [comment]
Ok, I see. I suppose adding options for changing defaults could (and probably should) be added later, on user requests.

**Comment by tolitius** [comment]
yea, changing defaults is _good to have_, can be done later.

just added: https://github.com/tolitius/mount#swapping-alternate-implementations so it makes it a bit harder to do `suspend`/`resume` given all the substitutions that will have 4 states instead of current 2.

but a bit harder is not impossible :)

**Comment by malchmih** [comment]
@tolitius What's the status of this? You need any help, PR or something else?

**Comment by tolitius** [comment]
@malchmih, just want to sit down and do it, but need at least one uninterrupted hour. #clojure_conj is almost over, if I am lucky I'll get that hour today. It _is_ coming :)

**Comment by malchmih** [comment]
Big thanks! =) I'd like to have it "in" to convert my existing project from component to mount, since I already use suspend/resume in a couple of components

**Comment by tolitius** [comment]
/_thinking almost out loud_

while it should not happen in the proper lifecycle flow, it will definitely happen in REPL or by mistake:

> if a state is `:stopped` and `(mount/resume)` is called, we should NOT start it, instead, we should throw and let developer know about the "flaw in the flow"

I do understand that it might complicate the `(reset)`, but it feels right. What do you think?

**Comment by malchmih** [comment]
Feels a bit strange to me, since in the original implementation only resume/suspend are getting called. I mean, you start in `:stopped` then call resume, then follows a sequence of `reset`s (which is `suspend` then `resume`).
But since you manage the whole state in the library, I guess it will be ok =)

**Comment by tolitius** [comment]
> you start in :stopped then call resume

just does not _sound_ right, which makes me think whether this is a correct behavior :)

I think the idea you brought up about suspend/resume is bigger than "just `tools.namespace`".
- It feels useful to be able to call `(mount/suspend)` from the REPL
- It could be used to suspend an application in production
- It could be used in between tests
- etc.

But I also see the other side where `resume` is a flavor of `start`.

So how about we WARN whenever a _suspendable state_ was stopped when `resume` was called. The state will still be started, but it won't be magical and rather explicitly clear that something was not right.

**Comment by malchmih** [comment]
Yeah, that's great

**Comment by tolitius** [comment]
on the other hand, I really don't like other components to stop on suspend and start on resume. it's not clean.

here is what I propose instead:
- `(mount/suspend)` will suspend _only_ suspendable states that are currently started
- `(mount/resume)` will resume _only_ suspendable states that are currently suspended
- `(mount/start)` will start all the states, except the ones that are currently _suspended_, it'll `resume` them instead
- `(mount/stop)` will stop everything

and here is what will enable `(reset)` to work
- `(mount/stop-except #app/state ...)` will stop all the states _except_ the ones in the list

this is clear. suspend/resume touch _only_ suspendable components, they don't start/stop anything, they just suspend and resume.

---

Can you try it in `suspended` branch and let me know whether it works for you?

I still have to do tests and docs, so I will formalize the `dev.clj` for suspendable states, but it would look similar to:

``` clojure
(defn start []
  (mount/start))  ;; will resume suspended states and start the rest

(defn stop []
  (mount/suspend) ;; which can also take a set of states if needed
  (mount/stop-except #'app/web-server #'app/db-connection))
```

all the `go`, `refresh`, etc.. will stay the same

**Comment by malchmih** [comment]
Seems fine to me =)
Sorry, won't have much next couple of days, will have to try it later. So don't wait for me on this.
Either way, implementation details are not _very_ critical for users (myself included). If it works in general, I'll make it work for my use case. Also :+1: for having tests

**Comment by tolitius** [comment]
this are _functional_/_behavior_ differences though :)

**Comment by tolitius** [comment]
done.

> in master

let me know after you try it

**Comment by tolitius** [comment]
@malchmih, let me know whether it works for you if/when you try it

**Comment by malchmih** [comment]
@tolitius Sorry it took me so long. I'm now moving my whole app from component to mount and it works perfectly! :blush: Already tested resume/suspend part, everything is fine. Thank you!

**Comment by tolitius** [comment]
great, thanks for the `suspendable` idea :)
